src/vsc_r_api.ts

```diff
@@ -0,0 +1,374 @@
+    // Is there a root folder open?
+
+    if (workspace.workspaceFolders.length === 1) {
+      // In single root common case, this will always work.
+      return {
+        path: workspace.workspaceFolders[0].uri.path
+      };
+    } else if (workspace.workspaceFolders.length > 1) {
+      // In less common multi-root folder case is a bit tricky. If the active
+      // text editor has scheme 'untitled:' (is unsaved), then
+      // workspace.getWorkspaceFolder() won't be able to find its Uri in any
+      // folder and will return undefined.
+      const currentDocument = getLastActiveTextEditor().document;
+      const currentDocFolder = workspace.getWorkspaceFolder(currentDocument.uri);
+      if (typeof currentDocFolder !== 'undefined') {
+        return {
+          path: currentDocFolder.uri.path
+        };
+      }
+    }
+  }
+
+  // if we got to here either: 
+  //   - the workspaceFolders array was undefined (no folder open)
+  //   - the activeText editor was an unsaved document, which has undefined workspace folder.
+  // return undefined and handle with a message in R.
+  return {
+    path: undefined
+  };
+}
+
+export async function documentNew(text: string, type: string, position: number[]) {
+  const documentUri = Uri.parse('untitled:' + path.join(projectPath().path, 'new_document.' + type));
+  const targetDocument = await workspace.openTextDocument(documentUri);
+  const edit = new WorkspaceEdit();
+  const docLines = targetDocument.lineCount;
+  edit.replace(documentUri,
+    targetDocument.validateRange(new Range(
+      new Position(0, 0),
+      new Position(docLines + 1, 0)
+    )),
+    text);
+
+  workspace.applyEdit(edit).then(async () => {
+    const editor = await window.showTextDocument(targetDocument);
+    editor.selections = [new Selection(
+      parsePosition(position, targetDocument),
+      parsePosition(position, targetDocument)
+    )];
+  });
+}
+
+// interface
+// represents addins in a QuickPick menu
+interface AddinItem extends QuickPickItem {
+  binding: string;
+  package: string;
+}
+
+let addinQuickPicks: AddinItem[] = undefined;
+
+export async function getAddinPickerItems() {
+
+  if (typeof addinQuickPicks === 'undefined') {
+    const addins: any[] = await fs.readJSON(path.join(sessionDir, 'addins.json'));
+    const addinItems = addins.map((x) => {
+      return {
+        alwaysShow: true,
+        description: `{${x.package}}`,
+        label: x.name,
+        detail: x.description,
+        picked: false,
+        binding: x.binding,
+        package: x.package,
+      };
+    });
+    addinQuickPicks = addinItems;
+  }
+  return addinQuickPicks;
+}
+
+export function purgeAddinPickerItems() {
+  addinQuickPicks = undefined;
+}
+
+export async function launchAddinPicker() {
+
+  if (!config().get<boolean>('sessionWatcher')) {
+    throw ('{rstudioapi} emulation requires session watcher to be enabled in extension config.');
+  }
+  if (typeof rTerm === 'undefined') {
```

Comment:
The addin picker still does not work with self-managed R sessions. It seems the check of `rTerm` here only applies to R sessions created by `R: Create R terminal` command.

We might need to check `sessionDir` instead of `rTerm` to ensure there's an attached session. We might also need to unset `sessionDir` when `rTerm` is closed.

Reply:
For example, check if sessionDir exists.

Reply:
tried this in 9b99786, seems to work right for me, but I don't have your advanced setup. I'd be keen to see what that looks like btw!